core/src/main/java/be/orbinson/aem/dictionarytranslator/services/impl/DictionaryServiceImpl.java

```diff
@@ -70,265 +52,209 @@
-                Privilege[] privileges = new Privilege[]{
-                        accessControlManager.privilegeFromName(Privilege.JCR_ADD_CHILD_NODES),
-                        accessControlManager.privilegeFromName(Privilege.JCR_REMOVE_NODE),
-                        accessControlManager.privilegeFromName("crx:replicate")
-                };
-                // https://jackrabbit.apache.org/oak/docs/nodestore/compositens.html#checking-for-read-only-access
-                boolean isPathWritable= session.hasCapability("addNode", path, new Object[] { "nt:folder" });
-                return isPathWritable && accessControlManager.hasPrivileges(path, privileges);
-            } catch (RepositoryException e) {
-                LOG.debug("Could not check if dictionary is editable, therefore assume it is not!", e);
-                return false;
+    private static final class LocaleComparator implements Comparator<Locale> {
+        @Override
+        public int compare(Locale o1, Locale o2) {
+            if (o1 == null && o2 == null) {
+                return 0;
             }
-        }
-        return false;
-    }
-
-    @Override
-    public int getOrdinal(Resource dictionaryResource) {
-        final String[] searchPaths = dictionaryResource.getResourceResolver().getSearchPath();
-        int i = 0;
-        for (; i < searchPaths.length; i++) {
-            if (dictionaryResource.getPath().startsWith(searchPaths[i])) {
-                return i;
+            if (o1 == null) {
+                return -1;
+            }
+            if (o2 == null) {
+                return 1;
             }
+            return o1.toLanguageTag().compareTo(o2.toLanguageTag());
         }
-        return i;
     }
 
-    public void addLanguage(Resource dictionary, String language, String basename) throws PersistenceException {
-        Map<String, Object> properties = new HashMap<>();
-
-        ResourceResolver resourceResolver = dictionary.getResourceResolver();
-
-        properties.put(ResourceResolver.PROPERTY_RESOURCE_TYPE, JcrResourceConstants.NT_SLING_FOLDER);
-        properties.put(JCR_LANGUAGE, language);
-        properties.put(JCR_MIXINTYPES, MIX_LANGUAGE);
-
-        if (StringUtils.isNotEmpty(basename)) {
-            properties.put(SLING_BASENAME, basename);
-        } else {
-            properties.put(SLING_BASENAME, dictionary.getPath());
+    @Override
+    public Collection<LanguageDictionary> getAllDictionaries(ResourceResolver resourceResolver) {
+        Iterator<Resource> iterator = resourceResolver
+                .findResources(QUERY_LANGUAGE_ROOTS, "xpath");
+        if (!allowEmptyQueryResults && !iterator.hasNext()) {
+            throw new IllegalStateException("The query '" + QUERY_LANGUAGE_ROOTS + "' returned no results. This either means the according search index is corrupt/non-existing or the underlying resource resolver does not support queries  (when executed from tests)!");
```

Comment:
is there a feature missing in sling mocks for this? I thought you could set pre-defined query results on a sling context? or ius this only for JCR based queries

Reply:
Fine by me to leave it, shouldn't happen in an OSGi context anyway